File: 알고리즘 연습/조합순열/Combination.py
# 라이브러리 사용
from itertools import combinations

result = list(combinations([1,2,3,4],2))
print(result)


# 직접 구현
arr = [1,2,3,4]
result = []
def comb(result, current, arr, start, k) :
    if k == 0 :
        
        result.append(current.copy()) # 깊은복사로 새로운 배열로 저장해야함
        return

    i = start
    while i <= len(arr)-k :
        current.append(arr[i])
        comb(result, current, arr, i+1, k-1)
        current.pop(-1)
        i += 1
    
comb(result, [], arr, 0, 2)
print(result)

File: 알고리즘 연습/조합순열/Combination.java
import java.util.ArrayList;
import java.util.List;

public class Combination {
    public static void main(String[] args) {
        int[] arr = {1,2,3,4};
        List<List<Integer>> result = new ArrayList<>();
        comb(result, new ArrayList<>(), arr, 0, 2);
        System.out.println(result);
    }

    // 배열 arr에서 k개의 값을 가지는 조합 추출
    static void comb(List<List<Integer>> result, List<Integer> current, int[] arr, int start, int k) {
        if (k == 0) {
            result.add(new ArrayList<>(current)); // 리스트안에 리스트
            return;
        }

        for (int i=start; i<=arr.length-k; i++) {
            current.add(arr[i]);
            comb(result, current, arr, i+1, k-1); // start를 0으로하면 중복가능 조합
            current.remove(current.size()-1);
        }
    }
}
